File: AdventOfCode/2018/AdventOfCode/Day07.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;

namespace AdventOfCode
{
    public static class Day07
    {
        public static string Part1(string input)
        {
            var regex = new Regex("Step ([A-Z]) must be finished before step ([A-Z]) can begin.", RegexOptions.Compiled);

            var steps = new HashSet<string>();
            var blockedBy = new Dictionary<string, List<string>>();

            foreach (var line in input.Split("\n".ToCharArray()))
            {
                var match = regex.Match(line);

                var first = match.Groups[1].ToString();
                var second = match.Groups[2].ToString();

                steps.Add(first);
                steps.Add(second);

                if (!blockedBy.ContainsKey(second))
                {
                    blockedBy[second] = new List<string>();
                }

                blockedBy[second].Add(first);
            }

            var result = new List<string>();
            var nodesToProcess = new List<string>();

            foreach (var step in steps)
            {
                if (!blockedBy.ContainsKey(step))
                {
                    nodesToProcess.Add(step);
                }
            }

            while (nodesToProcess.Count > 0)
            {
                var node = nodesToProcess.OrderBy(n => n).First();
                nodesToProcess.Remove(node);
                result.Add(node);

                var toRemove = new List<string>();

                foreach (var blocked in blockedBy)
                {
                    if (blocked.Value.Contains(node))
                    {
                        blocked.Value.Remove(node);

                        if (blocked.Value.Count == 0)
                        {
                            toRemove.Add(blocked.Key);
                            nodesToProcess.Add(blocked.Key);
                        }
                    }
                }

                foreach (var item in toRemove)
                {
                    blockedBy.Remove(item);
                }
            }

            return string.Join("", result);
        }

        public static int Part2(string input)
        {
            return -1;
        }
    }
}

File: AdventOfCode/2018/AdventOfCode/Day07.py
import re

from collections import defaultdict


def main():
    blocked_by = defaultdict(set)
    steps = set()

    for line in open("Day07.txt"):
        match = re.search("Step ([A-Z]) must be finished before step ([A-Z]) can begin.", line)
        blocked_by[match[2]].add(match[1])
        steps.add(match[1])
        steps.add(match[2])

    cur_tick = 0
    cur_processing = list()
    max_concurrent_processes = 5
    base_step_execution_time = 60

    while len(steps) > 0 or len(cur_processing) > 0:
        finished_items = list(filter(lambda x: x[1] <= cur_tick, cur_processing))

        for finished_item in finished_items:
            for item in blocked_by:
                if finished_item[0] in blocked_by[item]:
                    blocked_by[item].remove(finished_item[0])

        cur_processing = list(filter(lambda x: x[1] > cur_tick, cur_processing))

        available_items = sorted(list(filter(lambda x: len(blocked_by[x]) == 0, steps)))

        for available_item in available_items:
            if len(cur_processing) == max_concurrent_processes:
                break

            duration = ord(available_item) - ord("A") + 1
            cur_processing.append((available_item, cur_tick + base_step_execution_time + duration))
            steps.remove(available_item)

        cur_tick += 1

    print(cur_tick - 1)


if __name__ == "__main__":
    main()
